programs/polymer-prover/src/lib.rs

```
@@ -0,0 +1,244 @@
+    }
+
+    pub fn clear_proof_cache(ctx: Context<ClearProofCache>) -> Result<()> {
+        msg!("proof cache successfully cleared");
+        ctx.accounts.cache_account.cache.clear();
+        Ok(())
+    }
+
+    pub fn load_proof(ctx: Context<LoadProof>, proof_chunk: Vec<u8>) -> Result<()> {
+        ctx.accounts.cache_account.cache.extend(proof_chunk.iter());
+        Ok(())
+    }
+
+    pub fn validate_event(ctx: Context<ValidateEvent>) -> Result<()> {
+        // this is set by the owner/deployer during initialize()
+        let internal = &ctx.accounts.internal;
+
+        let result = validate_event::handler(
+            &ctx.accounts.cache_account.cache,
+            &internal.client_type,
+            &internal.signer_addr,
+            internal.peptide_chain_id,
+        );
+
+        msg!("{}", result);
+
+        // Determine if the current instruction is a Cross-Program Invocation (CPI)
+        // by comparing the instruction's program_id to the current program's ID.
+        // If they differ, the instruction was invoked by another program.
+        let is_cpi = {
+            let ix = ctx.accounts.instructions.to_account_info();
+            let index = load_current_index_checked(&ix)? as usize;
+            let current_ix = load_instruction_at_checked(index, &ix)?;
+            current_ix.program_id != *ctx.program_id
+        };
+
+        // if we are being called by another program, set the return data so they can pick it up
+        // otherwise (we are being called by an off-chain agent) emit an event for them to parse
+        if is_cpi {
+            set_return_data(&result.try_to_vec()?);
+        } else if let ValidateEventResult::Valid(chain_id, event) = result {
+            emit!(ValidateEventEvent {
+                chain_id,
+                topics: event.topics,
+                emitting_contract: *event.emitting_contract.as_bytes(),
+                unindexed_data: event.unindexed_data,
+            })
+        }
```

Validation failures in non-CPI calls should return an error; otherwise, the instruction succeeds even when validation fails.